client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornChannelBufferReader.java

```diff
@@ -0,0 +1,328 @@
+  private static final Logger LOG = LoggerFactory.getLogger(CelebornChannelBufferReader.class);
+
+  private CelebornChannelBufferManager bufferManager;
+
+  private final FlinkShuffleClientImpl client;
+
+  private final int shuffleId;
+
+  private final int partitionId;
+
+  private final TieredStorageInputChannelId inputChannelId;
+
+  private final int subPartitionIndexStart;
+
+  private final int subPartitionIndexEnd;
+
+  private final BiConsumer<ByteBuf, TieredStorageSubpartitionId> dataListener;
+
+  private final BiConsumer<Throwable, TieredStorageSubpartitionId> failureListener;
+
+  private final Consumer<RequestMessage> messageConsumer;
+
+  private CelebornBufferStream bufferStream;
+
+  private boolean isOpened;
+
+  private volatile boolean closed = false;
+
+  private volatile ConcurrentHashMap<Integer, Integer> subPartitionRequiredSegmentIds;
+
+  /** Note this field is to record the number of backlog before the read is set up. */
+  private int numBackLog = 0;
+
+  public CelebornChannelBufferReader(
+      FlinkShuffleClientImpl client,
+      ShuffleResourceDescriptor shuffleDescriptor,
+      TieredStorageInputChannelId inputChannelId,
+      int startSubIdx,
+      int endSubIdx,
+      BiConsumer<ByteBuf, TieredStorageSubpartitionId> dataListener,
+      BiConsumer<Throwable, TieredStorageSubpartitionId> failureListener) {
+    this.client = client;
+    this.shuffleId = shuffleDescriptor.getShuffleId();
+    this.partitionId = shuffleDescriptor.getPartitionId();
+    this.inputChannelId = inputChannelId;
+    this.subPartitionIndexStart = startSubIdx;
+    this.subPartitionIndexEnd = endSubIdx;
+    this.dataListener = dataListener;
+    this.failureListener = failureListener;
+    this.subPartitionRequiredSegmentIds = JavaUtils.newConcurrentHashMap();
+    for (int subPartitionId = subPartitionIndexStart;
+        subPartitionId <= subPartitionIndexEnd;
+        subPartitionId++) {
+      subPartitionRequiredSegmentIds.put(subPartitionId, -1);
+    }
+    this.messageConsumer =
+        requestMessage -> {
+          // Note that we need to use SubPartitionReadData because the isSegmentGranularityVisible
+          // is set as true when opening stream
+          if (requestMessage instanceof SubPartitionReadData) {
+            dataReceived((SubPartitionReadData) requestMessage);
+          } else if (requestMessage instanceof BacklogAnnouncement) {
+            backlogReceived(((BacklogAnnouncement) requestMessage).getBacklog());
+          } else if (requestMessage instanceof TransportableError) {
+            errorReceived(((TransportableError) requestMessage).getErrorMessage());
+          } else if (requestMessage instanceof BufferStreamEnd) {
+            onStreamEnd((BufferStreamEnd) requestMessage);
+          }
+        };
+  }
+
+  public void setup(TieredStorageMemoryManager memoryManager) {
+    this.bufferManager = new CelebornChannelBufferManager(memoryManager, this);
+    if (numBackLog > 0) {
```

Comment:
When setting up, why does this check `numBackLog`? Is there any case that `numBackLog` > 0?

Reply:
When the `bufferReader` is created but the setup method (called by the `InputGate`) is not invoked yet, it may receive `backLog` from the worker (as the stream is already opened). 

In this case, the `bufferReader` should accumulate the `backLog` and wait for the `setup` method to consume these `backLog`.